domain/removal/state/unit.go

```diff
@@ -123,6 +87,91 @@ AND    life_id = 0`, entityAssociationCount{})
 	return mUUID, nil
 }
 
+// markMachineAsDyingIfAllUnitsAreNotAlive checks if all the units on the
+// machine are not alive. If this is the case, it marks the machine as dying.
+func (st *State) markMachineAsDyingIfAllUnitsAreNotAlive(
+	ctx context.Context, tx *sqlair.TX, uUUID string,
+) (machineUUID string, err error) {
+	unitUUID := entityUUID{UUID: uUUID}
+
+	lastUnitStmt, err := st.Prepare(`
+WITH units_alive AS (
+    SELECT uuid, net_node_uuid
+    FROM   unit
+    WHERE  life_id = 0
+), units_not_alive AS (
+    SELECT uuid, net_node_uuid
+    FROM   unit
+    WHERE  life_id != 0
+), machines AS (
+    SELECT    m.uuid AS machine_uuid,
+              m.net_node_uuid,
+              COUNT(ua.uuid) AS unit_alive_count,
+              COUNT(una.uuid) AS unit_not_alive_count,
+			  COUNT(mp.parent_uuid) AS associated_count
+    FROM      machine AS m
+    JOIN      net_node AS nn ON nn.uuid = m.net_node_uuid
+    LEFT JOIN units_alive AS ua ON ua.net_node_uuid = nn.uuid
+    LEFT JOIN units_not_alive AS una ON una.net_node_uuid = nn.uuid
+	LEFT JOIN machine_parent AS mp ON mp.parent_uuid = m.uuid
+    GROUP BY  m.uuid
+)
+SELECT unit_alive_count AS &entityAssociationAliveCount.alive_count,
+       unit_not_alive_count AS &entityAssociationAliveCount.not_alive_count,
+	   associated_count AS &entityAssociationAliveCount.associated_count,
+       machine_uuid AS &entityAssociationAliveCount.uuid
+FROM   machines
+LEFT JOIN unit AS u ON u.net_node_uuid = machines.net_node_uuid
+WHERE  u.uuid = $entityUUID.uuid;
+    `, unitUUID, entityAssociationAliveCount{})
+	if err != nil {
+		return "", errors.Errorf("preparing unit count query: %w", err)
+	}
+
+	updateMachineStmt, err := st.Prepare(`
+UPDATE machine
+SET    life_id = 1
+WHERE  uuid = $entityUUID.uuid
+AND    life_id = 0`, entityUUID{})
+	if err != nil {
+		return "", errors.Errorf("preparing machine life update: %w", err)
+	}
+
+	var unitCount entityAssociationAliveCount
```

Comment:
`unitCount` is a bad variable name given the child machine count, and the type name isn't so good either as mentioned above.